x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/query/H3LatLonGeometry.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.spatial.index.query;
+
+import org.apache.lucene.geo.Component2D;
+import org.apache.lucene.geo.GeoUtils;
+import org.apache.lucene.geo.LatLonGeometry;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.spatial3d.geom.GeoArea;
+import org.apache.lucene.spatial3d.geom.GeoAreaFactory;
+import org.apache.lucene.spatial3d.geom.GeoPoint;
+import org.apache.lucene.spatial3d.geom.GeoPolygon;
+import org.apache.lucene.spatial3d.geom.GeoPolygonFactory;
+import org.apache.lucene.spatial3d.geom.PlanetModel;
+import org.elasticsearch.h3.CellBoundary;
+import org.elasticsearch.h3.H3;
+import org.elasticsearch.h3.LatLng;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/** Implementation of a lucene {@link LatLonGeometry} that covers the extent of a provided H3 bin. Note that
+ * H3 bin are polygons on the sphere. */
+class H3LatLonGeometry extends LatLonGeometry {
+
+    private final String h3Address;
+
+    H3LatLonGeometry(String h3Address) {
+        this.h3Address = h3Address;
+    }
+
+    @Override
+    protected Component2D toComponent2D() {
+        return new H3Polygon2D(h3Address);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o instanceof H3LatLonGeometry h3) {
+            return Objects.equals(h3Address, h3.h3Address);
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(h3Address);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("H3 : ");
+        sb.append("\"");
+        sb.append(h3Address);
+        sb.append("\"");
+        return sb.toString();
+    }
+
+    private static class H3Polygon2D implements Component2D {
+
+        private final long h3;
+        private final int res;
+        private final GeoPolygon hexagon;
+
+        private H3Polygon2D(String h3Address) {
+            h3 = H3.stringToH3(h3Address);
+            res = H3.getResolution(h3Address);
+            hexagon = getGeoPolygon(h3Address);
+            // I tried to compute the bounding box to set min/man values, but it seems to fail
+            // due to numerical errors. For now, we just don't use it, this means we will not be
+            // using the optimization provided by lucene's ComponentPredicate.
```

Comment:
I think the problem here is border effects because some of the edges of an H3 polygon do not belong to that geometry but to the neighbour. `ComponentPredicate` seems to break that and adding some buffer looked too hacky. We can try to work on that in another iteration, wdyt?

Reply:
Sounds good. I was worried about performance issues here, but this is a brand new feature, to at least there is no performance regression, and future performance improvements are always nice.